File: app/src/main/java/com/salah/tracker/data/calculator/PrayerTimeCalculator.kt
package com.salah.tracker.data.calculator

import java.time.LocalTime
import kotlin.math.*

class PrayerTimeCalculator {

    enum class CalculationMethod(val fajrAngle: Double, val ishaAngle: Double, val isTimeBasedIsha: Boolean, val ishaIntervalMinutes: Int) {
        MWL(18.0, 17.0, false, 0),
        ISNA(15.0, 15.0, false, 0),
        EGYPT(19.5, 17.5, false, 0),
        KARACHI(18.0, 18.0, false, 0),
        UMM_AL_QURA(18.5, 0.0, true, 90),
        GULF(19.5, 0.0, true, 90),
        TEHRAN(17.7, 14.0, false, 0)
    }

    enum class JuristicMethod(val shadowRatio: Double) {
        STANDARD(1.0), // Shafi'i, Maliki, Ja'fari, Hanbali
        HANAFI(2.0)
    }

    data class PrayerTimes(
        val fajr: LocalTime,
        val sunrise: LocalTime,
        val dhuhr: LocalTime,
        val asr: LocalTime,
        val maghrib: LocalTime,
        val isha: LocalTime
    )

    private fun dToR(deg: Double): Double = deg * Math.PI / 180.0
    private fun rToD(rad: Double): Double = rad * 180.0 / Math.PI

    private fun fixHour(h: Double): Double = (h % 24.0 + 24.0) % 24.0
    private fun fixAngle(a: Double): Double = (a % 360.0 + 360.0) % 360.0

    private fun getJulianDate(year: Int, month: Int, day: Int): Double {
        var y = year
        var m = month
        if (m <= 2) {
            y -= 1
            m += 12
        }
        val a = floor(y / 100.0)
        val b = 2.0 - a + floor(a / 4.0)
        return floor(365.25 * (y + 4716)) + floor(30.6001 * (m + 1)) + day + b - 1524.5
    }

    private data class SunPosition(val declination: Double, val equationOfTime: Double)

    private fun getSunPosition(jd: Double): SunPosition {
        val t = (jd - 2451545.0) / 36525.0
        
        // Solar mean longitude
        val l0 = fixAngle(280.46646 + 36000.76983 * t + 0.0003032 * t * t)
        
        // Solar mean anomaly
        val m = fixAngle(357.52911 + 35999.05029 * t - 0.0001537 * t * t)
        
        // Obliquity of the ecliptic
        val e0 = 23.439291 - 0.01300416 * t - 0.000000164 * t * t + 0.000000503 * t * t * t
        val ob = e0 + 0.00256 * cos(dToR(125.04 - 1934.136 * t))
        
        // Sun's equation of center
        val c = (1.914602 - 0.004817 * t - 0.000014 * t * t) * sin(dToR(m)) +
                (0.019993 - 0.000101 * t) * sin(dToR(2.0 * m)) +
                0.002893 * sin(dToR(3.0 * m))
        
        // Sun's true longitude
        val lambda = fixAngle(l0 + c)
        
        // Right ascension
        var ra = rToD(atan2(cos(dToR(ob)) * sin(dToR(lambda)), cos(dToR(lambda))))
        ra = fixAngle(ra)
        
        // Adjust RA to be in the same quadrant as lambda
        val lq = floor(lambda / 90.0) * 90.0
        val rq = floor(ra / 90.0) * 90.0
        val raAdjusted = ra + (lq - rq)
        
        // Sun's declination
        val dec = rToD(asin(sin(dToR(ob)) * sin(dToR(lambda))))
        
        // Equation of time (in hours)
        var diff = l0 - raAdjusted
        if (diff > 180.0) diff -= 360.0
        if (diff < -180.0) diff += 360.0
        val eqt = diff / 15.0
        
        return SunPosition(dec, eqt)
    }

    private fun getHourAngle(angle: Double, latitude: Double, declination: Double, isMorning: Boolean): Double {
        val latRad = dToR(latitude)
        val decRad = dToR(declination)
        val angleRad = dToR(angle)
        
        val cosH = (sin(-angleRad) - sin(latRad) * sin(decRad)) / (cos(latRad) * cos(decRad))
        // Clamp cosH to prevent NaN in extreme latitudes
        val clampedCosH = cosH.coerceIn(-1.0, 1.0)
        
        val h = rToD(acos(clampedCosH))
        return if (isMorning) -h else h
    }

    private fun getHourAngleForPositiveAltitude(altDeg: Double, latitude: Double, declination: Double): Double {
        val latRad = dToR(latitude)
        val decRad = dToR(declination)
        val altRad = dToR(altDeg)
        
        val cosH = (sin(altRad) - sin(latRad) * sin(decRad)) / (cos(latRad) * cos(decRad))
        val clampedCosH = cosH.coerceIn(-1.0, 1.0)
        
        return rToD(acos(clampedCosH))
    }

    fun calculate(
        latitude: Double,
        longitude: Double,
        timezoneOffset: Double,
        year: Int,
        month: Int,
        day: Int,
        method: CalculationMethod = CalculationMethod.MWL,
        juristic: JuristicMethod = JuristicMethod.STANDARD
    ): PrayerTimes {
        // Compute Julian date at noon (UTC)
        val jd = getJulianDate(year, month, day) - longitude / 360.0
        val sun = getSunPosition(jd)
        
        // Dhuhr (solar transit local time)
        // Midday (Dhuhr) is when sun reaches its highest transit point
        val dhuhrLocalHours = fixHour(12.0 - longitude / 15.0 - sun.equationOfTime + timezoneOffset)
        
        // Sunrise & Sunset (Maghrib)
        // Standard sunrise/sunset is 0.833 degrees below horizon (refraction + solar radius)
        val sunriseH = getHourAngle(0.833, latitude, sun.declination, isMorning = true)
        val sunsetH = getHourAngle(0.833, latitude, sun.declination, isMorning = false)
        
        val sunriseLocalHours = fixHour(dhuhrLocalHours + sunriseH / 15.0)
        val sunsetLocalHours = fixHour(dhuhrLocalHours + sunsetH / 15.0)
        
        // Fajr
        val fajrH = getHourAngle(method.fajrAngle, latitude, sun.declination, isMorning = true)
        val fajrLocalHours = fixHour(dhuhrLocalHours + fajrH / 15.0)
        
        // Asr shadow angle calculation
        val shadowLength = juristic.shadowRatio + tan(abs(dToR(latitude - sun.declination)))
        val altRad = atan(1.0 / shadowLength)
        val altDeg = rToD(altRad)
        val asrH = getHourAngleForPositiveAltitude(altDeg, latitude, sun.declination)
        val asrLocalHours = fixHour(dhuhrLocalHours + asrH / 15.0)
        
        // Isha
        val ishaLocalHours = if (method.isTimeBasedIsha) {
            // Isha is a fixed interval after Maghrib (e.g. 90 minutes)
            fixHour(sunsetLocalHours + method.ishaIntervalMinutes / 60.0)
        } else {
            val ishaH = getHourAngle(method.ishaAngle, latitude, sun.declination, isMorning = false)
            fixHour(dhuhrLocalHours + ishaH / 15.0)
        }

        // Add 1 minute safety buffer to Dhuhr
        val dhuhrFinal = fixHour(dhuhrLocalHours + 1.0 / 60.0)

        return PrayerTimes(
            fajr = doubleToTime(fajrLocalHours),
            sunrise = doubleToTime(sunriseLocalHours),
            dhuhr = doubleToTime(dhuhrFinal),
            asr = doubleToTime(asrLocalHours),
            maghrib = doubleToTime(sunsetLocalHours),
            isha = doubleToTime(ishaLocalHours)
        )
    }

    private fun doubleToTime(hoursDouble: Double): LocalTime {
        val totalMinutes = (hoursDouble * 60.0).roundToInt()
        val h = (totalMinutes / 60) % 24
        val m = totalMinutes % 60
        return LocalTime.of(h, m)
    }
}

File: flutter_app/lib/data/calculator/prayer_time_calculator.dart
import 'dart:math';
import 'package:flutter/material.dart';

enum CalculationMethod {
  mwl(18.0, 17.0, false, 0),
  isna(15.0, 15.0, false, 0),
  egypt(19.5, 17.5, false, 0),
  karachi(18.0, 18.0, false, 0),
  ummAlQura(18.5, 0.0, true, 90),
  gulf(19.5, 0.0, true, 90),
  tehran(17.7, 14.0, false, 0);

  final double fajrAngle;
  final double ishaAngle;
  final bool isTimeBasedIsha;
  final int ishaIntervalMinutes;

  const CalculationMethod(this.fajrAngle, this.ishaAngle, this.isTimeBasedIsha, this.ishaIntervalMinutes);
}

enum JuristicMethod {
  standard(1.0),
  hanafi(2.0);

  final double shadowRatio;

  const JuristicMethod(this.shadowRatio);
}

class PrayerTimes {
  final TimeOfDay fajr;
  final TimeOfDay sunrise;
  final TimeOfDay dhuhr;
  final TimeOfDay asr;
  final TimeOfDay maghrib;
  final TimeOfDay isha;

  PrayerTimes({
    required this.fajr,
    required this.sunrise,
    required this.dhuhr,
    required this.asr,
    required this.maghrib,
    required this.isha,
  });
}

class SunPosition {
  final double declination;
  final double equationOfTime;

  SunPosition(this.declination, this.equationOfTime);
}

class PrayerTimeCalculator {
  double _dToR(double deg) => deg * pi / 180.0;
  double _rToD(double rad) => rad * 180.0 / pi;

  double _fixHour(double h) => (h % 24.0 + 24.0) % 24.0;
  double _fixAngle(double a) => (a % 360.0 + 360.0) % 360.0;

  double _getJulianDate(int year, int month, int day) {
    int y = year;
    int m = month;
    if (m <= 2) {
      y -= 1;
      m += 12;
    }
    double a = (y / 100.0).floorToDouble();
    double b = 2.0 - a + (a / 4.0).floorToDouble();
    return (365.25 * (y + 4716)).floorToDouble() + (30.6001 * (m + 1)).floorToDouble() + day + b - 1524.5;
  }

  SunPosition _getSunPosition(double jd) {
    double t = (jd - 2451545.0) / 36525.0;
    
    double l0 = _fixAngle(280.46646 + 36000.76983 * t + 0.0003032 * t * t);
    double m = _fixAngle(357.52911 + 35999.05029 * t - 0.0001537 * t * t);
    
    double e0 = 23.439291 - 0.01300416 * t - 0.000000164 * t * t + 0.000000503 * t * t * t;
    double ob = e0 + 0.00256 * cos(_dToR(125.04 - 1934.136 * t));
    
    double c = (1.914602 - 0.004817 * t - 0.000014 * t * t) * sin(_dToR(m)) +
            (0.019993 - 0.000101 * t) * sin(_dToR(2.0 * m)) +
            0.002893 * sin(_dToR(3.0 * m));
            
    double lambda = _fixAngle(l0 + c);
    
    double ra = _rToD(atan2(cos(_dToR(ob)) * sin(_dToR(lambda)), cos(_dToR(lambda))));
    ra = _fixAngle(ra);
    
    double lq = (lambda / 90.0).floorToDouble() * 90.0;
    double rq = (ra / 90.0).floorToDouble() * 90.0;
    double raAdjusted = ra + (lq - rq);
    
    double dec = _rToD(asin(sin(_dToR(ob)) * sin(_dToR(lambda))));
    
    double diff = l0 - raAdjusted;
    if (diff > 180.0) diff -= 360.0;
    if (diff < -180.0) diff += 360.0;
    double eqt = diff / 15.0;
    
    return SunPosition(dec, eqt);
  }

  double _getHourAngle(double angle, double latitude, double declination, bool isMorning) {
    double latRad = _dToR(latitude);
    double decRad = _dToR(declination);
    double angleRad = _dToR(angle);
    
    double cosH = (sin(-angleRad) - sin(latRad) * sin(decRad)) / (cos(latRad) * cos(decRad));
    double clampedCosH = cosH.clamp(-1.0, 1.0);
    
    double h = _rToD(acos(clampedCosH));
    return isMorning ? -h : h;
  }

  double _getHourAngleForPositiveAltitude(double altDeg, double latitude, double declination) {
    double latRad = _dToR(latitude);
    double decRad = _dToR(declination);
    double altRad = _dToR(altDeg);
    
    double cosH = (sin(altRad) - sin(latRad) * sin(decRad)) / (cos(latRad) * cos(decRad));
    double clampedCosH = cosH.clamp(-1.0, 1.0);
    
    return _rToD(acos(clampedCosH));
  }

  PrayerTimes calculate({
    required double latitude,
    required double longitude,
    required double timezoneOffset,
    required int year,
    required int month,
    required int day,
    CalculationMethod method = CalculationMethod.mwl,
    JuristicMethod juristic = JuristicMethod.standard,
  }) {
    double jd = _getJulianDate(year, month, day) - longitude / 360.0;
    SunPosition sun = _getSunPosition(jd);
    
    double dhuhrLocalHours = _fixHour(12.0 - longitude / 15.0 - sun.equationOfTime + timezoneOffset);
    
    double sunriseH = _getHourAngle(0.833, latitude, sun.declination, true);
    double sunsetH = _getHourAngle(0.833, latitude, sun.declination, false);
    
    double sunriseLocalHours = _fixHour(dhuhrLocalHours + sunriseH / 15.0);
    double sunsetLocalHours = _fixHour(dhuhrLocalHours + sunsetH / 15.0);
    
    double fajrH = _getHourAngle(method.fajrAngle, latitude, sun.declination, true);
    double fajrLocalHours = _fixHour(dhuhrLocalHours + fajrH / 15.0);
    
    double shadowLength = juristic.shadowRatio + tan((_dToR(latitude - sun.declination)).abs());
    double altRad = atan(1.0 / shadowLength);
    double altDeg = _rToD(altRad);
    double asrH = _getHourAngleForPositiveAltitude(altDeg, latitude, sun.declination);
    double asrLocalHours = _fixHour(dhuhrLocalHours + asrH / 15.0);
    
    double ishaLocalHours;
    if (method.isTimeBasedIsha) {
      ishaLocalHours = _fixHour(sunsetLocalHours + method.ishaIntervalMinutes / 60.0);
    } else {
      double ishaH = _getHourAngle(method.ishaAngle, latitude, sun.declination, false);
      ishaLocalHours = _fixHour(dhuhrLocalHours + ishaH / 15.0);
    }

    double dhuhrFinal = _fixHour(dhuhrLocalHours + 1.0 / 60.0);

    return PrayerTimes(
      fajr: _doubleToTime(fajrLocalHours),
      sunrise: _doubleToTime(sunriseLocalHours),
      dhuhr: _doubleToTime(dhuhrFinal),
      asr: _doubleToTime(asrLocalHours),
      maghrib: _doubleToTime(sunsetLocalHours),
      isha: _doubleToTime(ishaLocalHours),
    );
  }

  TimeOfDay _doubleToTime(double hoursDouble) {
    int totalMinutes = (hoursDouble * 60.0).round();
    int h = (totalMinutes ~/ 60) % 24;
    int m = totalMinutes % 60;
    return TimeOfDay(hour: h, minute: m);
  }
}
